### ISSUE TITLE
[purs ide] Add a declaration type Filter

### ISSUE THREAD

**Issue by kRITZCREEK** [purescript/purescript#2924]
Add a filter which allows to filter for the type of declaration.

## Usecases

I'll list a couple of example use cases that would use this feature.

### Add type class instance

An editor could provide an `add type class instance` function to the user which inserted a snippet like:

`instance ?t?c :: ?c ?t where`

which would then prompt for a completion for the class name, filtered to only contain type class names. It would then use that completion to fill in the `?c` slots. Next it could check whether that type class is defined in the current module. If it isn't:

Ask the user for a type completion, with types defined in the current module (importantly we don't want type synonyms here) and use the result to fill in the `?t` slot.

If the type class is defined in the current module, just prompt for all the types defined in the current project and add an import on selection (filter the import addition down to the type declaration right away, so we don't end up with data constructors or kinds with the same name).

## Add deriving clauses
Very similar to `add type class`

## Adding more specific filters

Filtering options that are specific to some kind of declaration could live within the declaration type filters. For example, a filter which only keeps types of a certain kind, or a filter which only keeps values that mention a certain type in their signature, etc.

# Implementation

To not have to break the format as soon as we start adding more declaration specific filters I propose to not just use a non-empty array of strings for the declaration types, but to wrap them in objects right away. Something like:

```json
{
  "filter": "declaration",
  "params": [
    { 
      "declarationtype": "synonym"
    },
    {
      "declarationtype": "value"
    },
    {
      "declarationtype": "typeclass"
    }
  ]
}
```

As this filter is strictly more general than the namespace filter, I propose we implement the namespace filter in terms of the declaration filter.

This is a good follow-up to the namespace filter so I'd reserve this one for @stefanholzmueller or @sectore, if you want it.

**Comment by stefanholzmueller** [comment]
Thanks for mentioning me, but I do not have enough spare time, at least in the next two weeks. So I'm fine if someone else picks this up. :)

**Comment by sectore** [comment]
@kRITZCREEK Nice, I will take over this part. Thx!

**Comment by kRITZCREEK** [comment]
Fixed in #2934
